# Project HoloMotion
#
# Copyright (c) 2024-2025 Horizon Robotics. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
# implied. See the License for the specific language governing
# permissions and limitations under the License.

network_interface: "eth0" # on Orin
device: "cuda"
policy_freq: 50 # Hz
control_freq: 200 # Hz
control_decimation: 4
msg_type: "hg" # "hg" or "go"
imu_type: "pelvis" # "torso" or "pelvis"
lowcmd_topic: "/lowcmd"
lowstate_topic: "/lowstate"
action_topic: "/humanoid/action"


policy_path: "your_model.onnx"
motion_file: "squat2stand.pkl" # Stand up
motion_file_3: "stand2squat.pkl" # squat down

motion_file_4: "0-BMLmovi_Subject_48_F_MoSh_Subject_48_F_10_stageii_padded.pkl" # up  jog in place
motion_file_6: "0-CMU_94_94_16_stageii_sliced-156-510_padded.pkl"  # right  stretch your body
motion_file_5: "0-HumanEva_S1_Box_1_stageii_padded.pkl"  # down practice martial arts
motion_file_7: "0-KIT_572_bow_slight03_stageii_padded.pkl"  # left bow

motion_duration: 24.63333333
context_length: 25
num_fut_frames: 1
num_bodies_extend: 25
num_dofs: 21
num_actions: 21
action_scale: 0.25
obs_scales:
  base_ang_vel: 0.25
  projected_gravity: 1.0
  dof_pos: 1.0
  dof_vel: 0.05
  actions: 1.0
  ref_motion_phase: 1.0
  history_actor: 1.0
  ref_dof_pos_flat: 1.0
  ref_dof_vel_flat: 1.0
  ref_motion_state_flat: 1.0

normalization:
  obs_clip: 100.0
  action_clip: 100.0

current_obs:
  - base_ang_vel
  - projected_gravity
  - dof_pos
  - dof_vel
  - actions
  - history_actor
  - ref_motion_state_flat
  # - ref_dof_pos_flat 
  # - ref_dof_vel_flat

history_obs:
  # hist are organized as [context_length x hist_1 , context_length x hist_2, ...]
  actions: 25
  base_ang_vel: 25
  dof_pos: 25
  dof_vel: 25
  projected_gravity: 25

obs_dims:
  base_ang_vel: 3
  projected_gravity: 3
  dof_pos: 21
  dof_vel: 21
  actions: 21
  ref_motion_phase: 1
  ref_motion_state_flat: 426
  # ref_dof_pos_flat: 210
  # ref_dof_vel_flat: 210

complete_dof_order:
  - left_hip_pitch_joint
  - left_hip_roll_joint
  - left_hip_yaw_joint
  - left_knee_joint
  - left_ankle_pitch_joint
  - left_ankle_roll_joint
  - right_hip_pitch_joint
  - right_hip_roll_joint
  - right_hip_yaw_joint
  - right_knee_joint
  - right_ankle_pitch_joint
  - right_ankle_roll_joint
  - waist_yaw_joint
  - waist_roll_joint
  - waist_pitch_joint
  - left_shoulder_pitch_joint
  - left_shoulder_roll_joint
  - left_shoulder_yaw_joint
  - left_elbow_joint
  - left_wrist_roll_joint
  - left_wrist_pitch_joint
  - left_wrist_yaw_joint
  - right_shoulder_pitch_joint
  - right_shoulder_roll_joint
  - right_shoulder_yaw_joint
  - right_elbow_joint
  - right_wrist_roll_joint
  - right_wrist_pitch_joint
  - right_wrist_yaw_joint

policy_dof_order:
  - left_hip_pitch_joint
  - left_hip_roll_joint
  - left_hip_yaw_joint
  - left_knee_joint
  - left_ankle_pitch_joint
  - left_ankle_roll_joint

  - right_hip_pitch_joint
  - right_hip_roll_joint
  - right_hip_yaw_joint
  - right_knee_joint
  - right_ankle_pitch_joint
  - right_ankle_roll_joint

  - waist_yaw_joint

  - left_shoulder_pitch_joint
  - left_shoulder_roll_joint
  - left_shoulder_yaw_joint
  - left_elbow_joint

  - right_shoulder_pitch_joint
  - right_shoulder_roll_joint
  - right_shoulder_yaw_joint
  - right_elbow_joint

dof2motor_idx_mapping:
  # https://support.unitree.com/home/zh/G1_developer/about_G1
  left_hip_pitch_joint: 0
  left_hip_roll_joint: 1
  left_hip_yaw_joint: 2
  left_knee_joint: 3
  left_ankle_pitch_joint: 4
  left_ankle_roll_joint: 5
  right_hip_pitch_joint: 6
  right_hip_roll_joint: 7
  right_hip_yaw_joint: 8
  right_knee_joint: 9
  right_ankle_pitch_joint: 10
  right_ankle_roll_joint: 11
  waist_yaw_joint: 12
  waist_roll_joint: 13
  waist_pitch_joint: 14
  left_shoulder_pitch_joint: 15
  left_shoulder_roll_joint: 16
  left_shoulder_yaw_joint: 17
  left_elbow_joint: 18
  left_wrist_roll_joint: 19
  left_wrist_pitch_joint: 20
  left_wrist_yaw_joint: 21
  right_shoulder_pitch_joint: 22
  right_shoulder_roll_joint: 23
  right_shoulder_yaw_joint: 24
  right_elbow_joint: 25
  right_wrist_roll_joint: 26
  right_wrist_pitch_joint: 27
  right_wrist_yaw_joint: 28

control_params:
  stiffness:
    left_hip_pitch_joint: 100.0
    left_hip_roll_joint: 100.0
    left_hip_yaw_joint: 100.0
    left_knee_joint: 200.0
    left_ankle_pitch_joint: 20.0
    left_ankle_roll_joint: 20.0

    right_hip_pitch_joint: 100.0
    right_hip_roll_joint: 100.0
    right_hip_yaw_joint: 100.0
    right_knee_joint: 200.0
    right_ankle_pitch_joint: 20.0
    right_ankle_roll_joint: 20.0

    waist_yaw_joint: 400.0

    left_shoulder_pitch_joint: 90.0
    left_shoulder_roll_joint: 60.0
    left_shoulder_yaw_joint: 20.0
    left_elbow_joint: 60.0
    left_wrist_roll_joint: 10.0
    left_wrist_pitch_joint: 10.0
    left_wrist_yaw_joint: 10.0

    right_shoulder_pitch_joint: 90.0
    right_shoulder_roll_joint: 60.0
    right_shoulder_yaw_joint: 20.0
    right_elbow_joint: 60.0
    right_wrist_roll_joint: 10.0
    right_wrist_pitch_joint: 10.0
    right_wrist_yaw_joint: 10.0

  damping:
    left_hip_pitch_joint: 2.5
    left_hip_roll_joint: 2.5
    left_hip_yaw_joint: 2.5
    left_knee_joint: 5.0
    left_ankle_pitch_joint: 0.2
    left_ankle_roll_joint: 0.1

    right_hip_pitch_joint: 2.5
    right_hip_roll_joint: 2.5
    right_hip_yaw_joint: 2.5
    right_knee_joint: 5.0
    right_ankle_pitch_joint: 0.2
    right_ankle_roll_joint: 0.1

    waist_yaw_joint: 5.0

    left_shoulder_pitch_joint: 2.0
    left_shoulder_roll_joint: 1.0
    left_shoulder_yaw_joint: 0.4
    left_elbow_joint: 1.0
    left_wrist_roll_joint: 0.1
    left_wrist_pitch_joint: 0.1
    left_wrist_yaw_joint: 0.1

    right_shoulder_pitch_joint: 2.0
    right_shoulder_roll_joint: 1.0
    right_shoulder_yaw_joint: 0.4
    right_elbow_joint: 1.0
    right_wrist_roll_joint: 0.1
    right_wrist_pitch_joint: 0.1
    right_wrist_yaw_joint: 0.1

# zhanli
default_joint_angles:
  left_hip_pitch_joint: -0.1
  left_hip_roll_joint: 0.0
  left_hip_yaw_joint: 0.0
  left_knee_joint: 0.3
  left_ankle_pitch_joint: -0.2
  left_ankle_roll_joint: 0.0

  right_hip_pitch_joint: -0.1
  right_hip_roll_joint: 0.0
  right_hip_yaw_joint: 0.0
  right_knee_joint: 0.3
  right_ankle_pitch_joint: -0.2
  right_ankle_roll_joint: 0.0

  waist_yaw_joint: 0.0
  waist_roll_joint: 0.0
  waist_pitch_joint: 0.0

  left_shoulder_pitch_joint: 0.0
  left_shoulder_roll_joint: 0.0
  left_shoulder_yaw_joint: 0.0
  left_elbow_joint: 0.0
  left_wrist_roll_joint: 0.0
  left_wrist_pitch_joint: 0.0
  left_wrist_yaw_joint: 0.0

  right_shoulder_pitch_joint: 0.0
  right_shoulder_roll_joint: 0.0
  right_shoulder_yaw_joint: 0.0
  right_elbow_joint: 0.0
  right_wrist_roll_joint: 0.0
  right_wrist_pitch_joint: 0.0
  right_wrist_yaw_joint: 0.0

# dun
default_joint_angles_start:
  left_hip_pitch_joint: -2.5
  left_hip_roll_joint: 0.01
  left_hip_yaw_joint: 0.02
  left_knee_joint: 2.9
  left_ankle_pitch_joint: -0.86
  left_ankle_roll_joint: 0.0

  right_hip_pitch_joint: -2.5
  right_hip_roll_joint: -0.01
  right_hip_yaw_joint: 0.02
  right_knee_joint: 2.9
  right_ankle_pitch_joint: -0.86
  right_ankle_roll_joint: 0.0

  waist_yaw_joint: -0.01
  waist_roll_joint: 0.0
  waist_pitch_joint: 0.0

  left_shoulder_pitch_joint: -0.25
  left_shoulder_roll_joint: 0.37
  left_shoulder_yaw_joint: 0.16
  left_elbow_joint: 1.08
  left_wrist_roll_joint: 0.0
  left_wrist_pitch_joint: 0.0
  left_wrist_yaw_joint: 0.0

  right_shoulder_pitch_joint: -0.23
  right_shoulder_roll_joint: -0.52
  right_shoulder_yaw_joint: -0.04
  right_elbow_joint: 1.28
  right_wrist_roll_joint: -0.0
  right_wrist_pitch_joint: -0.0
  right_wrist_yaw_joint: 0.0

default_joint_angles_stand:
  left_hip_pitch_joint: -0.1
  left_hip_roll_joint: 0.0
  left_hip_yaw_joint: 0.0
  left_knee_joint: 0.3
  left_ankle_pitch_joint: -0.2
  left_ankle_roll_joint: 0.0

  right_hip_pitch_joint: -0.1
  right_hip_roll_joint: 0.0
  right_hip_yaw_joint: 0.0
  right_knee_joint: 0.3
  right_ankle_pitch_joint: -0.2
  right_ankle_roll_joint: 0.0

  waist_yaw_joint: 0.0
  waist_roll_joint: 0.0
  waist_pitch_joint: 0.0

  left_shoulder_pitch_joint: 0.0
  left_shoulder_roll_joint: 0.0
  left_shoulder_yaw_joint: 0.0
  left_elbow_joint: 0.0
  left_wrist_roll_joint: 0.0
  left_wrist_pitch_joint: 0.0
  left_wrist_yaw_joint: 0.0

  right_shoulder_pitch_joint: 0.0
  right_shoulder_roll_joint: 0.0
  right_shoulder_yaw_joint: 0.0
  right_elbow_joint: 0.0
  right_wrist_roll_joint: 0.0
  right_wrist_pitch_joint: 0.0
  right_wrist_yaw_joint: 0.0

# pd
joint_limits:
  position:
    left_hip_pitch_joint: [-2.5307, 2.8798]
    left_hip_roll_joint: [-0.5236, 2.9671]
    left_hip_yaw_joint: [-2.7576, 2.7576]
    left_knee_joint: [-0.087267, 2.8798]
    left_ankle_pitch_joint: [-0.87267, 0.5236]
    left_ankle_roll_joint: [-0.2618, 0.2618]
    right_hip_pitch_joint: [-2.5307, 2.8798]
    right_hip_roll_joint: [-2.9671, 0.5236]
    right_hip_yaw_joint: [-2.7576, 2.7576]
    right_knee_joint: [-0.087267, 2.8798]
    right_ankle_pitch_joint: [-0.87267, 0.5236]
    right_ankle_roll_joint: [-0.2618, 0.2618]
    waist_yaw_joint: [-2.618, 2.618]
    left_shoulder_pitch_joint: [-3.0892, 2.6704]
    left_shoulder_roll_joint: [-1.5882, 2.2515]
    left_shoulder_yaw_joint: [-2.618, 2.618]
    left_elbow_joint: [-1.0472, 2.0944]
    right_shoulder_pitch_joint: [-3.0892, 2.6704]
    right_shoulder_roll_joint: [-2.2515, 1.5882]
    right_shoulder_yaw_joint: [-2.618, 2.618]
    right_elbow_joint: [-1.0472, 2.0944]

  velocity:
    left_hip_pitch_joint: 32
    left_hip_roll_joint: 20
    left_hip_yaw_joint: 32
    left_knee_joint: 20
    left_ankle_pitch_joint: 37
    left_ankle_roll_joint: 37
    right_hip_pitch_joint: 32
    right_hip_roll_joint: 20
    right_hip_yaw_joint: 32
    right_knee_joint: 20
    right_ankle_pitch_joint: 37
    right_ankle_roll_joint: 37
    waist_yaw_joint: 32
    left_shoulder_pitch_joint: 37
    left_shoulder_roll_joint: 37
    left_shoulder_yaw_joint: 37
    left_elbow_joint: 37
    right_shoulder_pitch_joint: 37
    right_shoulder_roll_joint: 37
    right_shoulder_yaw_joint: 37
    right_elbow_joint: 37

  effort:
    left_hip_pitch_joint: 88
    left_hip_roll_joint: 139
    left_hip_yaw_joint: 88
    left_knee_joint: 139
    left_ankle_pitch_joint: 50
    left_ankle_roll_joint: 50
    right_hip_pitch_joint: 88
    right_hip_roll_joint: 139
    right_hip_yaw_joint: 88
    right_knee_joint: 139
    right_ankle_pitch_joint: 50
    right_ankle_roll_joint: 50
    waist_yaw_joint: 88
    left_shoulder_pitch_joint: 25
    left_shoulder_roll_joint: 25
    left_shoulder_yaw_joint: 25
    left_elbow_joint: 25
    right_shoulder_pitch_joint: 25
    right_shoulder_roll_joint: 25
    right_shoulder_yaw_joint: 25
    right_elbow_joint: 25
limit_scales:
  position: 1.5 # Allows 50% more range of motion
  velocity: 1.2 
  effort: 1.5 
